### PR TITLE
Update page header and add divider line and scroll animation

### LINKED ISSUE #1155: [Theme] Refactor/Update of Page Header
Collection of tasks to enhance the Page Header Toolbar.

TO DO list:

- [x] Investigate the structure of currrent header (Page - Header - Toolbar - back btn - btn's)
- [x] #2073
- [x] Header/toolbar title bold (#1455)
- [x] Enable option for subtitle on page (#811)
- [x] Dropshadow on header/toolbar (in Scroll state) (#2040)
- [x] #1981
- [x] #2219 
- [x] #2074
- [x] Check all styles and size of header/toolbar
- [x] Make it possible to trigger moving of content from the page area to the toolbar in various steps (added by @MadsBuchmann)
- [x] #2493 
- [ ] Make and update issues for the above mentioned tasks

**Describe the solution you'd like**

Header : toolbar : Page

https://zpl.io/bzz03KM
https://zpl.io/aXRgj8M

### PR DESCRIPTION
**Pull request by mictro**
## Which issue does this PR close?

This PR closes #1155 

## What is the new behavior?

Add the following changes to the Kirby page header.
o Header/toolbar title in bold.
o Add option for subtitle on page.
o Add dynamic toolbar height.
o Add dropshadow on header/toolbar when scrolled.
o Correct all styles and size of header/toolbar.
o Change page actions to always show in toolbar.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Checklist:

The following tasks should be carried out in sequence in order to follow [the process of contributing](https://github.com/kirbydesign/designsystem/blob/main/.github/CONTRIBUTING.md/#the-process-of-contributing) correctly.

### Reminders
- [x] Make sure you have implemented tests following the guidelines in: "[The good: Test](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Test)".
- [x] Make sure you have updated the cookbook with examples and showcases (for bug fixes, enhancements & new components).

### Review  
- [x] Do a [self-review](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Self-review).
- [x] Request that the changes are code-reviewed 
- [x] Request that the changes are [UX reviewed](https://github.com/kirbydesign/designsystem/blob/main/.github/CONTRIBUTING.md/#ux-review) (only necessary if your PR introduces visual changes)

When the pull request has been approved it will be merged to `develop` by Team Kirby.

### CODE REVIEW

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -1 +1 @@
-*.scss  @jkaltoft
+*.scss  @goofy
```
Comment (RasmusKjeldgaard) [inline .github/CODEOWNERS]: What is going on here? 🤔 
Should probably be reverted 😄

File: apps/cookbook/src/app/examples/page-example/advanced/page-advanced-example.component.ts [hunk 2]
```diff
@@ -18,7 +18,7 @@ const config = {
 
   <!-- Custom Page Subtitle -->
   <div *kirbyPageSubtitle>
-<span>Custom subtitle with a long long long long name</span>
+    Custom subtitle with a long long long long name
```
Comment (RasmusKjeldgaard) [inline apps/cookbook/src/app/examples/page-example/advanced/page-advanced-example.component.ts:21]: Just out of curiosity - was this just removed because it was unnecessary clutter, or is there any other reason?

File: libs/designsystem/src/lib/components/fab-sheet/fab-sheet.component.ts [hunk 3]
```diff
@@ -52,7 +52,7 @@ export class FabSheetComponent implements AfterContentInit {
   ) {}
 
   ngAfterContentInit(): void {
-    if (!!this.actionSheet) {
+    if (this.actionSheet) {
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/src/lib/components/fab-sheet/fab-sheet.component.ts]: This looks like a linting auto correction that is unrelated to all the other changes.

File: libs/designsystem/src/lib/components/page/page.component.scss [hunk 4]
```diff
@@ -1,6 +1,10 @@
 @use '@kirbydesign/core/src/scss/interaction-state';
 @use '@kirbydesign/core/src/scss/utils';
 
+:host {
+  display: contents;
+}
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/src/lib/components/page/page.component.scss]: Could we add a note about why this somewhat unusual `display: contents` property is needed? 

I am pretty sure we discussed this as a nice alternative to doing all the `removeWrapper()` logic in `page.component.ts` in the original PR, but would be great to have a small comment here on that.

File: libs/designsystem/src/lib/components/page/page.component.scss [hunk 5]
```diff
@@ -34,11 +80,24 @@ ion-toolbar {
   }
 
   ion-title {
+    // Animation on opacity and a vertical translation with 10px gives visually pleasing experience
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/src/lib/components/page/page.component.scss]: I think you are completely right. 
But I guess this is kinda subjective and maybe the comment can just be removed?

File: libs/designsystem/src/lib/components/page/page.component.scss [hunk 6]
```diff
@@ -173,7 +240,7 @@ ion-content {
  */
 .sticky-content-container {
   position: sticky;
-  top: 0;
+  top: -#{utils.size('xs')}; // Adjust for padding-top of ion-content to position static content properly.
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/src/lib/components/page/page.component.scss]: This has the potential of breaking if the padding of ion-content is changed down the line. 
Would be great if the sticky content could be 'self-contained' - can we think of any other way of achieving this, e.g. adjust ion-content itself?

File: libs/designsystem/src/lib/components/page/page.component.scss [hunk 7]
```diff
@@ -182,4 +249,22 @@ ion-content {
     max-width: var(--page-content-max-width, utils.$page-content-max-width);
     margin: 0 auto;
   }
+
+  &.content-sticking {
+    // Adjust for inine padding of ion-content to stretch divider to full width.
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/src/lib/components/page/page.component.scss]: smol typo `inine`

File: libs/designsystem/src/lib/components/page/page.component.scss [hunk 8]
```diff
@@ -182,4 +249,22 @@ ion-content {
...
+    margin-inline-start: calc(-1 * var(--page-content-padding-start, #{utils.size('s')}));
+    margin-inline-end: calc(-1 * var(--page-content-padding-end, #{utils.size('s')}));
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/src/lib/components/page/page.component.scss]: Would it not be better to apply this calculation to the divider in isolation (:before just below)
```
left: calc(-1 * var(--page-content-padding-start, #{utils.size('s')}));
right: calc(-1 * var(--page-content-padding-end, #{utils.size('s')}));
```

File: libs/designsystem/src/lib/components/page/page.component.scss [hunk 9]
```diff
@@ -182,4 +249,22 @@ ion-content {
...
+
+    // Divider
+    &::before {
+      content: '';
+      background-color: utils.get-color('medium');
+      position: absolute;
+      height: 1px;
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/src/lib/components/page/page.component.scss]: This is more of a UX concern, but the 1px height in combination with the color makes the divider quite hard to see. 
If I squint even a little it disappears so maybe we should push back a little accessibility-wise? Either a darker color or larger height might fix this.

File: libs/designsystem/src/lib/components/page/page.component.scss [hunk 10]
```diff
@@ -182,4 +249,22 @@ ion-content {
...
+      bottom: 0;
+      z-index: 1;
+      left: 0;
+      width: 100%;
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/src/lib/components/page/page.component.scss]: width can be removed if you implement the left/right suggestion.

File: libs/designsystem/src/lib/components/page/page.component.ts [hunk 11]
```diff
@@ -239,34 +244,32 @@ export class PageComponent
...
   private isActive: boolean;
 
   private ngOnDestroy$: Subject<void> = new Subject<void>();
+  private contentScrolled$: Observable<ScrollDetail>;
+
   private navigationStart$: Observable<RouterEvent> = this.router.events.pipe(
-    takeUntil(this.ngOnDestroy$),
-    filter((event: RouterEvent) => event instanceof NavigationStart)
+    filter((event: RouterEvent) => event instanceof NavigationStart),
+    takeUntil(this.ngOnDestroy$)
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/src/lib/components/page/page.component.ts:257]: Awesome, thats one more ✅  for the performance audit!

image

File: libs/designsystem/src/lib/components/page/page.component.ts [hunk 12]
```diff
@@ -281,6 +284,17 @@ export class PageComponent
...
   ngAfterViewInit(): void {
+    this.contentScrolled$ = this.content.ionScroll.pipe(
+      debounceTime(contentScrollDebounceTimeInMS),
+      map((event) => event.detail),
+      takeUntil(this.ngOnDestroy$)
+    );
+
+    this.contentScrolled$.subscribe((scrollInfo: ScrollDetail) => {
+      this.isContentScrolled = scrollInfo.scrollTop > contentScrolledOffsetInPixels;
+      this.changeDetectorRef.detectChanges();
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/src/lib/components/page/page.component.ts:294]: Are we aware of any possible performance issues with this?
If I read it correctly, we might get quite a lot of change detection cycles when scrolling.

File: libs/designsystem/src/lib/components/page/page.component.html [hunk 13]
```diff
@@ -66,7 +58,8 @@
 
   <!--  Sticky content -->
   <ng-container *ngIf="stickyContentTemplate">
-    <div class="sticky-content-container">
+    <span #stickyContentContainerPreamble></span>
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/src/lib/components/page/page.component.html]: Are we adding this span strictly for the purpose of observing it with the intersection observer? 
Could we do it something like the following instead of introducing a new element: https://css-tricks.com/how-to-detect-when-a-sticky-element-gets-pinned/

**Review by RasmusKjeldgaard — CHANGES_REQUESTED** [review]
Great work! I have a mixed bag of questions and suggestions for you 😄

File: apps/cookbook/src/app/examples/page-example/advanced/page-advanced-example.component.ts (hunk 2 above)
Comment (mictro) [reply apps/cookbook/src/app/examples/page-example/advanced/page-advanced-example.component.ts:21]: Yes, was removed during review of #1455 due to unnecessary mark-up :)

File: .github/CODEOWNERS (hunk 1 above)
Comment (mictro) [reply .github/CODEOWNERS]: Done 😄

File: libs/designsystem/src/lib/components/fab-sheet/fab-sheet.component.ts (hunk 3 above)
Comment (mictro) [reply libs/designsystem/src/lib/components/fab-sheet/fab-sheet.component.ts]: Yes, apparently the linter is setup to correct these - I.M.O. the original is somewhat clearer type-wise (see https://airbnb.io/javascript/#coercion--booleans)

Comment (RasmusKjeldgaard) [reply libs/designsystem/src/lib/components/fab-sheet/fab-sheet.component.ts]: My concern was more about the change being included here, when the file has nothing to do with the feature. We could consider to revert it.

File: libs/designsystem/src/lib/components/page/page.component.ts (hunk 11 above)
Comment (mictro) [reply libs/designsystem/src/lib/components/page/page.component.ts:257]: Yup, follows https://ncjamieson.com/avoiding-takeuntil-leaks/

File: libs/designsystem/src/lib/components/page/page.component.ts (hunk 12 above)
Comment (mictro) [reply libs/designsystem/src/lib/components/page/page.component.ts:294]: No, the event stream is throttled by a debouce of 10ms so only very few event gets through :)

File: libs/designsystem/src/lib/components/page/page.component.scss (hunk 5 above)
Comment (mictro) [reply libs/designsystem/src/lib/components/page/page.component.scss]: Agree, removed

File: libs/designsystem/src/lib/components/page/page.component.scss (hunk 10 above)
Comment (mictro) [reply libs/designsystem/src/lib/components/page/page.component.scss]: Done

File: libs/designsystem/src/lib/components/page/page.component.scss (hunk 7 above)
Comment (mictro) [reply libs/designsystem/src/lib/components/page/page.component.scss]: Typo fixed :)

File: libs/designsystem/src/lib/components/page/page.component.scss (hunk 8 above)
Comment (mictro) [reply libs/designsystem/src/lib/components/page/page.component.scss]: Yes indeed, implemented!

File: libs/designsystem/src/lib/components/page/page.component.scss (hunk 6 above)
Comment (mictro) [reply libs/designsystem/src/lib/components/page/page.component.scss]: Done

File: libs/designsystem/src/lib/components/page/page.component.html (hunk 13 above)
Comment (mictro) [reply libs/designsystem/src/lib/components/page/page.component.html]: Implemented the css solution.

File: libs/designsystem/src/lib/components/page/page.component.scss (hunk 9 above)
Comment (mictro) [reply libs/designsystem/src/lib/components/page/page.component.scss]: The default divider has been aligned with the local-nav divider above and the new change has been approved by UX @henrikvoetmand

**Review by RasmusKjeldgaard — APPROVED** [review]
Looks beautiful!

Consider a more descriptive title for the PR, maybe something along the lines of:
"Add divider line and scroll animation to page header"
